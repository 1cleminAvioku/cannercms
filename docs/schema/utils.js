export const galleryValidation = {
  validator: (content, reject) => {
    if (content.length === 0) {
      return reject("should at least have one photo");
    }
  }
};