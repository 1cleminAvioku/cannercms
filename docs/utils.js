/* globals FIREBASE_API_KEY */

import {FirebaseClientService} from '@canner/image-service-config';
import firebase from 'firebase';
import {FirebaseRtdbClientConnector, PrismaClient} from 'canner-graphql-interface';


firebase.initializeApp({
  apiKey: FIREBASE_API_KEY,
  authDomain: "test-new-qa.firebaseapp.com",
  databaseURL: "https://test-new-qa.firebaseio.com",
  projectId: "test-new-qa",
  storageBucket: "test-new-qa.appspot.com",
  messagingSenderId: "983887338585"
});

const defaultApp = firebase.app();
const connector = new FirebaseRtdbClientConnector({
  database: defaultApp.database()
});


const storage = new FirebaseClientService({
  firebase: firebase,
  dir: "the/path/to", // specify the path you want upload to 
  filename: "filename", // rename file without extension
  hash: true, // if true, the filename will add a hash string, e.g.: `filename-${hash}.jpg`
}).getServiceConfig();

export default {
  storage,
  connector,
  graphClient: new PrismaClient()
}
