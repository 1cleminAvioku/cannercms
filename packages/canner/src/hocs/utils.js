// @flow
import {mapValues, get, isPlainObject, isArray} from 'lodash';

export function getValue(value: any, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (isPlainObject(result)) {
      if ('edges' in result && 'pageInfo' in result) {
        return get(result, ['edges', key, 'node']);
      }
      return get(result, key);
    } else if (isArray(result)) {
      return get(result, key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: any) {
  if (isPlainObject(value)) {
    if (value.edges && value.pageInfo) {
      return value.edges.map(edge => parseConnectionToNormal(edge.node));
    }
    return mapValues(value, item => parseConnectionToNormal(item));
  } else if (isArray(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

export function defaultValue(type: string, relation: any) {
  switch (type) {
    case 'connection': {
      return {
        edges: [],
        pageInfo: {
          hasNextPage: false,
          hasPreviousPage: false
        }
      }
    }
    case 'array': {
      return [];
    }
    case 'object': {
      return {};
    }
    case 'boolean': {
      return false;
    }
    case 'number': {
      return 0;
    }
    case 'string': {
      return '';
    }
    case 'relation': {
      if (relation.type === 'toMany') {
        return {
          edges: [],
          pageInfo: {
            hasNextPage: false,
            hasPreviousPage: false
          }
        };
      } else {
        return null;
      }
    }
    case 'image':
    case 'file': {
      return {
        url: '',
        contentType: '',
        name: '',
        size: 0,
        __typename: null
      }
    }
    case 'geoPoint': {
      return {
        placeId: '',
        address: '',
        lat: 122,
        lng: 23
      };
    }
    default: {
      return null;
    }
  }
}