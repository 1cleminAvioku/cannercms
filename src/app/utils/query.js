/**
 * @flow
 */
type dataType = {
    totalPage: number,
    data: Array<any>
  } | Array<any>

import getByPath from 'lodash/get';

export default class Query {
  data: dataType

  constructor({data}: {data: Array<any>}) {
    this.data = data;
  }

  filter(filter: {[key: string]: any} = {}) {
    const data = this.getPureData();
    this.setData(Object.keys(filter).reduce((result, key) => {
      return this._filterByKey(result, key, filter[key]);
    }, data));
    return this;
  }

  sort(sort: {[key: string]: 1 | -1} = {}) {
    const data = this.getPureData();
    // eslint-disable-next-line max-len
    const collator = new Intl.Collator(undefined, {numeric: true, sensitivity: 'base', caseFirst: 'lower'});
    const key = Object.keys(sort)[0];
    const order = sort[key];
    this.setData(data.sort((a, b) => {
    // eslint-disable-next-line max-len
      return order > 0 ? collator.compare(a[key], b[key]) : collator.compare(b[key], a[key]);
    }));
    return this;
  }

  static turnPage(pagination?: {start: number, limit: number}) {
    if (!pagination) {
      return {page: 1, perPage: 10};
    }
    const {start, limit} = pagination;
    return {
      page: Math.ceil(start / limit) + 1,
      perPage: limit,
    };
  }

  page(pagination?: {page: number, perPage: number}) {
    const data = this.getPureData();
    if (!pagination) {
      return this;
    }
    const {page = 1, perPage = 10} = pagination;
    const queryData = data.slice(perPage * (page - 1), perPage * page);
    this.setData({
      goTo: (page) => ({
        start: (page - 1) * perPage,
        limit: perPage,
      }),
      page,
      totalPage: Math.ceil(data.length / perPage) || 0,
      data: queryData,
    });
    return this;
  }

  getPureData(): Array<any> {
    if (this.data && 'data' in this.data) {
      return (this.data: any).data;
    }
    return (this: any).data;
  }

  setData(data: dataType) {
    if (this.data && this.data.data) {
      (this.data: any).data = data;
    } else {
      this.data = data;
    }
  }

  getData() {
    return this.data;
  }

  // eslint-disable-next-line max-len
  _filterByKey(data: Array<any>, key: string, conditions: {[key: string]: any}) {
    return data.filter((datum) => {
      // eslint-disable-next-line max-len
      return Object.keys(conditions).reduce((result, conditionkey) => {
        // eslint-disable-next-line max-len
        return result && isValid(getByPath(datum, key.split('/'), ''), conditionkey, conditions[conditionkey]);
      }, true);
    });
  }
}

export function isValid(value: any, condition: string, conditionValue: any) {
  switch (condition) {
    case '$eq':
    case '$equal':
      // eslint-disable-next-line eqeqeq
      return value == conditionValue;
    case '$gt':
      return value > conditionValue;
    case '$gte':
      return value >= conditionValue;
    case '$lt':
      return value < conditionValue;
    case '$lte':
      return value <= conditionValue;
    case '$contains':
      return value.indexOf && value.indexOf(conditionValue) !== -1;
    case '$in':
      return conditionValue.indexOf && conditionValue.indexOf(value) !== -1;
    case '$regex':
      return new RegExp(conditionValue, 'g').test(value);
    default:
      return true;
  }
}

// eslint-disable-next-line max-len
export function passQuery(action: MutateAction, queryKey: string, totalPage: number) {
  if (action.type !== 'CREATE_ARRAY_ITEM') {
    return true;
  }
  // 用來判斷 CollectionUpadte 是否要在這個 query 的 data 下 mutate
  const allPage = queryKey === 'all';
  const query = allPage ? {filter: {}, pagination: {}} : JSON.parse(queryKey);
  const lastPage = getByPath(query, 'pagination.page', 1) === totalPage;
  const dataIsValid = new Query({data: [action.payload.value.toJS()]})
    .filter(query.filter).getPureData().length > 0;
  return dataIsValid && (allPage || lastPage);
}
