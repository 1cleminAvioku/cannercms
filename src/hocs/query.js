// @flow

import * as React from 'react';
import {Spin, Icon} from 'antd';
import type RefId from 'canner-ref-id';
import {Map, List, is} from 'immutable';
import Toolbar from './components/toolbar';
import type {Query} from '../query';
import {mapValues} from 'lodash';

const antIcon = <Icon type="loading" style={{fontSize: 24}} spin />;

type Props = {
  type: string,
  items: Object,
  refId: RefId,
  query: Query,
  fetch: FetchDef,
  subscribe: SubscribeDef,
  updateQuery: Function,
  ui: string,
  path: string,
  pattern: string,
  relation: {
    type: string,
    to: string
  },
  toolbar: {
    sort?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    pagination?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    filter?: {
      component?: React.ComponentType<*>,
      [string]: *
    },
    toolbarLayout?: {
      component?: React.ComponentType<*>,
      [string]: *
    }
  }
};

type State = {
  value: any,
  rootValue: any,
  originRootValue: any,
  isFetching: boolean,
}

export default function withQuery(Com: React.ComponentType<*>) {
  // this hoc will fetch data;
  return class ComponentWithQuery extends React.PureComponent<Props, State> {
    key: string;
    subscription: any;

    constructor(props: Props) {
      super(props);
      this.state = {
        value: null,
        rootValue: null,
        originRootValue: null,
        isFetching: true
      };
      this.key = props.refId.getPathArr()[0];
    }

    componentDidMount() {
      this.queryData();
    }

    UNSAFE_componentWillReceiveProps(props: Props) {
      const {refId} = this.props;
      
      if (refId.toString() !== props.refId.toString()) {
        this.queryData(props);
      } 
    }

    componentWillUnmount() {
      this.unsubscribe();
    }

    getRootValue = () => {
      return this.state.rootValue;
    }

    queryData = (props?: Props): Promise<*> => {
      const {refId, fetch} = props || this.props;
      return fetch(this.key).then(data => {
        this.setState({
          originRootValue: data,
          rootValue: parseConnectionToNormal(data),
          value: getValue(data, refId.getPathArr()),
          isFetching: false
        });
        this.subscribe();
      });
    }

    unsubscribe = () => {
      if (this.subscription) {
        this.subscription.unsubscribe();
      }
    }

    subscribe = () => {
      const {subscribe, refId} = this.props;
      const subscription = subscribe(this.key, (newOriginRootValue) => {
        const {originRootValue} = this.state
        if (shouldUpdate(originRootValue, newOriginRootValue)) {
          this.setState({
            originRootValue: newOriginRootValue,
            rootValue: parseConnectionToNormal(newOriginRootValue),
            value: getValue(newOriginRootValue, refId.getPathArr()),
          });
        }
      });
      this.subscription = subscription;
    }

    render() {
      const {value, isFetching, rootValue} = this.state;
      const {toolbar, query, refId, items, type, path, relation, updateQuery, pattern} = this.props;
      if (isFetching) {
        return <Spin indicator={antIcon} />;
      }
      if (pattern === 'array' || (type === 'relation' && relation.type === 'toMany')) {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v]);
        return <Toolbar items={items} toolbar={toolbar} args={args} query={query} refId={refId} value={value} updateQuery={updateQuery}>
          <Com {...this.props} rootValue={rootValue} value={value.getIn(['edges'], new List()).map(item => item.get('node'))} />
        </Toolbar>;
      } else if (type === 'relation' && relation.type === 'toOne') {
        const queries = query.getQueries(path.split('/')).args || {pagination: {first: 10}};
        const variables = query.getVairables();
        const args = mapValues(queries, v => variables[v]);
        return <Toolbar items={items} toolbar={toolbar} args={args} query={query} refId={refId} value={value} updateQuery={updateQuery}>
          <Com {...this.props} rootValue={rootValue} value={value || new Map()} />
        </Toolbar>;
      }
      return <Com {...this.props} rootValue={rootValue} value={value} />;
    }
  };
}

export function getValue(value: Map<string, *>, idPathArr: Array<string>) {
  return idPathArr.reduce((result: any, key: string) => {
    if (Map.isMap(result)) {
      if (result.has('edges') && result.has('pageInfo')) {
        return result.getIn(['edges', key, 'node']);
      }
      return result.get(key);

    } else if (List.isList(result)) {
      return result.get(key);
    } else {
      return result;
    }
  }, value);
}

export function parseConnectionToNormal(value: Map<string, *> | List<*>) {
  if (Map.isMap(value)) {
    value = ((value: any): Map<string, any>);
    if (value.has('edges') && value.has('pageInfo')) {
      return (value.get('edges'): any).map(edge => parseConnectionToNormal(edge.get('node')));
    }
    return value.map(item => parseConnectionToNormal(item));
  } else if (List.isList(value)) {
    return value.map(item => parseConnectionToNormal(item))
  } else {
    return value;
  }
}

function shouldUpdate(value: any, newValue: any) {
  return !is(value, newValue);
}