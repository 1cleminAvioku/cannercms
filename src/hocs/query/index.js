// @flow

import * as React from 'react';
import Sort from './sort';
import Pagination from './pagination';
import Filter from './filter';
import isString from 'lodash/isString';
import isUndefined from 'lodash/isUndefined';
import type RefId from 'canner-ref-id';

type Props = {
  filter: filterType,
  sort: sortType,
  params: Object,
  routes: Array<string>,
  componentId: string,
  fetch: FetchDef,
  refId: RefId
}

type State = {
  filter: filterType,
  sort: sortType,
  pagination: {
    start: number,
    limit: number,
  },
  paginationRes: any,
  value: any,
}

export default (Plugin: React.ComponentType<*>) => {
  return class PluginWithQuery extends React.PureComponent<Props, State> {
    constructor(props: Props) {
      super(props);
      this.state = {
        filter: {},
        sort: {},
        pagination: {
          start: 0,
          limit: 10,
        },
        paginationRes: {},
        value: null,
      };
      this.changeFilter = this.changeFilter.bind(this);
      this.changeSort = this.changeSort.bind(this);
      this.changePagination = this.changePagination.bind(this);
    }

    componentWillMount() {
      const {refId, subscribe, componentId} = this.props;
      this.queryData()
        .then(() => {
          subscribe(refId.toString(), componentId, 'value', (value) => {
            this.setState({
              value,
            });
          });
        });
    }

    queryData() {
      const {refId, fetch, componentId} = this.props;
      const {filter, sort, pagination} = this.state;
      return fetch(refId.toString(), componentId, {filter, sort, pagination}).then((ctx) => {
        this.setState({
          paginationRes: ctx.response.pagination,
          value: ctx.response.body,
        });
      });
    }

    changeFilter(condition) {
      this.setState({
        filter: condition,
      }, this.queryData);
    }

    changeSort(condition) {
      this.setState({
        sort: condition,
      }, this.queryData);
    }

    changePagination(start, limit) {
      this.setState({
        pagination: {
          start,
          limit,
        },
      }, this.queryData);
    }

    render() {
      const {filter, sort, params, routes} = this.props;
      const {paginationRes, pagination, value} = this.state;
      const op = params && params.op;
      if (!value) {
        return null;
      }
      return (
        <div>
          {
            // 如果是 string 則不是交給這個 filter 處理
            // 是 restful-qa 會在 render 前 filter
            // 目前只有 ireconcile 使用到
            op === 'create' || routes.length > 1 || isString(filter) || isUndefined(filter) ?
              null :
              <Filter onChange={this.changeFilter} schema={filter}/>
          }
          {
            op === 'create' || routes.length > 1 || isUndefined(sort) ?
              null :
              <Sort onChange={this.changeSort} options={sort}/>
          }
          <Plugin {...this.props} query={{filter: this.state.filter, sort: this.state.sort, pagination}} rootValue={value} value={value} showPagination={false}/>
          {
            op === 'create' || routes.length > 1 ?
              null :
              <Pagination onChange={this.changePagination} pagination={pagination} paginationRes={paginationRes}/>
          }
        </div>
      );
    }
  };
};

