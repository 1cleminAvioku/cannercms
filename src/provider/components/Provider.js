/**
 * @flow
 */

import * as React from 'react';
import PropTypes from 'prop-types';
import App from '../app';
import {Bucket, Cache, EndpointMiddleware, Store} from '../middleware';
import type Endpoint from '../endpoint';
import isArray from 'lodash/isArray';
import isEqual from 'lodash/isEqual';

type Props = {
  schema: {[key: string]: any},
  endpoint: {[key: string]: Endpoint},
  dataDidChange: void => void,
  children: React.ChildrenArray<React.Node>
}

type State = {

}

export default class Provider extends React.PureComponent<Props, State> {
  app: App

  static childContextTypes = {
    subscribe: PropTypes.func,
    request: PropTypes.func,
    fetch: PropTypes.func,
    deploy: PropTypes.func,
  };

  constructor(props: Props) {
    super(props);
    (this: any).request = this.request.bind(this);
    (this: any).subscribe = this.subscribe.bind(this);
    (this: any).fetch = this.fetch.bind(this);
    const {schema, endpoint} = props;
    this.app = new App()
      .use(new Store())
      .use(new Bucket())
      .use(new Cache())
      .use(new EndpointMiddleware({schema, endpoint}));
  }

  componentWillReceiveProps(nextProps: Props) {
    const {schema} = this.props;
    const schemaKey = Object.keys(schema);
    const nextSchemaKey = Object.keys(nextProps.schema);
    if (!isEqual(schemaKey, nextSchemaKey)) {
      this.app = new App()
        .use(new Store())
        .use(new Bucket())
        .use(new Cache())
        .use(new EndpointMiddleware({schema: nextProps.schema, endpoint: nextProps.endpoint}));
    }
  }

  getChildContext() {
    return {
      subscribe: this.subscribe,
      request: this.request,
      fetch: this.fetch,
      deploy: this.deploy,
      // createAction
    };
  }

  fetch(key: string, componentId: string, query: queryType, mutate: Mutate): Promise<*> {
    return this.app.handleChange({
      request: {
        type: 'fetch',
        key,
        query,
        componentId,
      },
      response: {
        mutate,
      },
    });
  }

  subscribe(key: string, componentId: string, subjectType: SubjectType, observer: rxjs$Observer<*>) {
    return this.app.handleChange({
      request: {
        type: 'subscribe',
        key,
        observer,
        componentId,
        subjectType,
      },
    }).then((ctx) => ctx.response.subscription);
  }

  request(action: MutateAction) {
    // const {cannerJSON} = this.state;
    const {dataDidChange} = this.props;
    if (action.type !== 'NOOP') {
      const {key} = action.payload;
      return this.app.handleChange({
        request: {
          type: 'write',
          action,
          key,
        },
      }).then(dataDidChange);
    }
  }

  deploy = (key: string, id: string) => {
    return this.app.handleChange({
      request: {
        type: 'deploy',
        key,
        id,
      },
    });
  }

  _splitKey(path: string | [string, string]) {
    return isArray(path) ?
      path[0].split('/')[0] :
      // $FlowFixMe
      path.split('/')[0];
  }

  render() {
    return React.Children.only(this.props.children);
  }
}
