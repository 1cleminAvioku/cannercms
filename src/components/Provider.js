/**
 * @flow
 */

import * as React from 'react';
import {HOCContext} from '../hocs/context';
import {ApolloProvider} from 'react-apollo';
import type ApolloClient from 'apollo-boost';
import isEmpty from 'lodash/isEmpty';
import {ActionManager, actionToMutation, actionsToVariables, mutatePure} from '../action';
import {Query} from '../query';
import type {Action, ActionType} from '../action/types';
import gql from 'graphql-tag';
import {fromJS} from 'immutable';
import {objectToQueries} from '../query/utils';
import mapValues from 'lodash/mapValues';

type Props = {
  schema: {[key: string]: any},
  dataDidChange: void => void,
  children: React.ChildrenArray<React.Node>,
  client: ApolloClient,
  rootKey: string
}

type State = {

}

export default class Provider extends React.PureComponent<Props, State> {
  actionManager: ActionManager;
  query: Query;
  observableQueryMap: {[string]: any}

  constructor(props: Props) {
    super(props);
    this.actionManager = new ActionManager();
    this.query = new Query({schema: props.schema});
    this.observableQueryMap = mapValues(props.schema, (v, key) => {
      return props.client.watchQuery({
        query: gql`${this.query.toGQL(key)}`
      });
    });
  }

  updateQuery = (path: string, args: Object) => {
    const paths = path.split('/');
    this.query.updateQueries(paths, 'args', args);
    this.observableQueryMap[paths[0]].setOptions({
      query: gql`${this.query.toGQL(paths[0])}`
    });
  }

  // path: posts/name args: {where, pagination, sort}
  fetch = (key: string): Promise.resolve<*> => {
    const observabale = this.observableQueryMap[key];
    const result = observabale.currentResult();

    return result.loading ?
      observabale.result()
        .then(result => {
          this.log('fetch', key, result);
          return fromJS(result.data);
        }) :
      Promise.resolve(result.data)
        .then(data => {
          this.log('fetch', key, result);
          return fromJS(data);
        });
  }

  subscribe = (key: string, callback: (data: any) => void) => {
    const observableQuery = this.observableQueryMap[key];
    return observableQuery.subscribe({
      next: () => {
        const {loading, errors, data} = observableQuery.currentResult();
        if (!loading && !errors && data && !isEmpty(data)) {
          // this.log('subscribe', key, data);
          callback(fromJS(data));
        }
      } 
    });
  }

  deploy = (key: string, id?: string): Promise.resolve<*> => {
    const {client} = this.props;
    let actions = this.actionManager.getActions(key, id);
    if (!actions || !actions.length) {
      return Promise.resolve();
    }
    
    actions = removeIdInCreateArray(actions);
    const mutation = actionToMutation(actions[0]);
    const variables = actionsToVariables(actions);
    return client.mutate({
      mutation: gql`${objectToQueries(mutation, false)}`,
      variables,
    }).then(result => {
      this.log('deploy', key, {
        id,
        variables,
        result
      });
      this.actionManager.removeActions(key, id);
      client.resetStore();
      return fromJS(result.data);
    });
  }

  reset = (key: string, id?: string) => {
    const {client} = this.props;
    this.actionManager.removeActions(key, id);
    return client.resetStore();
  }

  request = (action: Action<ActionType>, options: {write: boolean} = {write: true}): Promise<*> => {
    const {client} = this.props;
    const {write = true} = options;
    this.actionManager.addAction(action);
    const query = gql`${this.query.toGQL(action.payload.key)}`;
    const data = client.readQuery({query});
    this.log('request', action, mutatePure(data, action), data);
    if (write) {
      client.writeQuery({
        query: query,
        data: mutatePure(data, action)
      });
    }
    return Promise.resolve();
  }

  log(type: string, ...payload: any) {
    if (process.env.NODE_ENV !== 'development') {
      return;
    }
    let color = "black";

    switch (type) {
      case "request":
        color = "Green";
        break;
      case "fetch":
        color = "DodgerBlue";
        break;
      case "deploy":
        color = "Red";
        break;
      case "subscribe":
        color = "Orange";
        break;
      default:
        break;
    }
    // eslint-disable-next-line
    console.log("%c" + type, "color:" + color, ...payload);
}

  render() {
    const {client} = this.props;
    return <ApolloProvider client={client}>
      <HOCContext.Provider value={{
        request: this.request,
        deploy: this.deploy,
        fetch: this.fetch,
        reset: this.reset,
        updateQuery: this.updateQuery,
        subscribe: this.subscribe,
        query: this.query
      }}>
        {React.Children.only(this.props.children)}
      </HOCContext.Provider>
    </ApolloProvider>
  }
}

function removeIdInCreateArray(actions: Array<Action<ActionType>>) {
  return actions.map(action => {
    if (action.type === 'CREATE_ARRAY') {
      action.payload.value = action.payload.value.update(v => {
        v = v.delete('id');
        v = v.delete('__typename');
        return v;
      });
    }
    return action;
  });
}
