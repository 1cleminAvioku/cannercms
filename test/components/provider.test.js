import * as React from 'react';
import Enzyme, { shallow } from 'enzyme';
// import toJson from 'enzyme-to-json';
import Adapter from '../react163Adapter';
import {schema, defaultData} from './data';
import Provider from '../../src/components/Provider';
// import {HOCContext as Context} from '../../src/hocs/context';
import {createClient} from '@canner/graphql-resolver';
import {schemaToQueriesObject, objectToQueries} from '../../src/query/utils';
import {fromJS} from 'immutable';
import gql from "graphql-tag";
import mapValues from 'lodash/mapValues';
import pick from 'lodash/pick';
Enzyme.configure({ adapter: new Adapter() });


function toGQL(schema, key) {
  return gql`${objectToQueries(pick(schemaToQueriesObject(schema), key))}`
}

describe('provider methods', () => {
  let client;
  beforeEach(() => {
    client = createClient({
      schema: mapValues(schema, v => {
        if (v.type === 'array') {
          return {
            type: 'array',
            items: v.items.items
          };
        }
        return v;
      }),
      defaultData
    })
  });

  test('fetch object', () => {
    const wrapper = shallow(<Provider
      schema={schema}
      client={client}
    >
      <div>children</div>
    </Provider>);
    return wrapper.instance().fetch('home').then(d => {
      expect(d.data).toMatchSnapshot();
    })
  });

  test('fetch array', () => {
    const wrapper = shallow(<Provider
      schema={schema}
      client={client}
    >
      <div>children</div>
    </Provider>)
    return wrapper.instance().fetch('posts').then(d => {
      expect(d.data).toMatchSnapshot();
    })
  });

  test('updateQuery', async () => {
    const wrapper = shallow(<Provider
      schema={schema}
      client={client}
    >
      <div>children</div>
    </Provider>)
    const originData = await wrapper.instance().fetch('posts');
    wrapper.instance().updateQuery('posts', {where: {author: {name: "user2"}}});
    const newData = await wrapper.instance().fetch('posts');
    expect(originData).not.toEqual(newData);
  });

  test('request', async () => {
    const wrapper = shallow(<Provider
      schema={schema}
      client={client}
    >
      <div>children</div>
    </Provider>);
    const instance = wrapper.instance();
    const action = {
      type: 'UPDATE_OBJECT',
      payload: {
        key: 'home',
        value: fromJS({
          count: 11,
        })
      }
    };
    await instance.fetch('home');
    instance.request(action);
    const newData = client.readQuery({
      query: toGQL(schema, 'home')
    });
    expect(newData.home.count).toBe(11);
  });

  test('request without writing', async () => {
    const wrapper = shallow(<Provider
      schema={schema}
      client={client}
    >
      <div>children</div>
    </Provider>);
    const instance = wrapper.instance();
    const action = {
      type: 'UPDATE_OBJECT',
      payload: {
        key: 'home',
        value: fromJS({
          count: 11,
        })
      }
    };
    await instance.fetch('home');
    instance.request(action, {write: false});
    const newData = client.readQuery({
      query: toGQL(schema, 'home')
    });
    expect(newData.home.count).toBe(10);
  });

  test('subscribe', async () => {
    const wrapper = shallow(<Provider
      schema={schema}
      client={client}
    >
      <div>children</div>
    </Provider>);
    const mockCallback = jest.fn();
    const instance = wrapper.instance();
    const action = {
      type: 'UPDATE_OBJECT',
      payload: {
        key: 'home',
        value: fromJS({
          count: 11,
        })
      }
    };
    await instance.fetch('home');
    const subscription = instance.subscribe('home', mockCallback);
    expect(subscription).toHaveProperty('unsubscribe');
    instance.request(action);
    return new Promise((resolver) => {
      setTimeout(resolver, 100);
    }).then(() => {
      expect(mockCallback.mock.calls[0][0]).toHaveProperty('home', expect.anything(Object));
    });
  });

  test('deploy', async () => {
    const wrapper = shallow(<Provider
      schema={schema}
      client={client}
    >
      <div>children</div>
    </Provider>);
    const instance = wrapper.instance();
    const action = {
      type: 'UPDATE_OBJECT',
      payload: {
        key: 'home',
        value: fromJS({
          count: 11,
        })
      }
    };
    await instance.fetch('home');
    instance.request(action, {write: false});
    await expect(instance.deploy('home')).resolves.toMatchObject({updateHome: {count: 11}});
    // const newData = client.readQuery({
    //   query: toGQL(schema, 'home')
    // });
    // expect(newData).toMatchObject({home: {
    //   count: 11
    // }});
  });
});