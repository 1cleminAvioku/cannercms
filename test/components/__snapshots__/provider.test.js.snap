// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`provider methods fetch array 1`] = `
Object {
  "posts": Object {
    "__typename": null,
    "edges": Array [],
    "pageInfo": Object {
      "__typename": null,
      "hasNextPage": false,
    },
  },
}
`;

exports[`provider methods fetch object 1`] = `
Object {
  "home": Object {
    "__typename": "HomePayload",
    "bestAuthor": Object {
      "__typename": "User",
      "age": 10,
      "email": "wwwy3y3@gmail.com",
      "id": "1",
      "images": Array [
        Object {
          "id": "0",
          "url": "url",
        },
      ],
      "name": "user1",
      "posts": Array [
        Object {
          "__typename": "Post",
          "author": "1",
          "id": "1",
          "notes": Array [
            Object {
              "text": "note1",
            },
            Object {
              "text": "note2",
            },
          ],
          "title": "123",
        },
        Object {
          "__typename": "Post",
          "author": "2",
          "id": "2",
          "notes": Array [
            Object {
              "text": "note3",
            },
            Object {
              "text": "note4",
            },
          ],
          "title": "123",
        },
      ],
    },
    "count": 10,
    "header": Object {
      "__typename": null,
      "subTitle": "subTitle",
      "title": "largeTitle",
    },
    "navs": Object {
      "__typename": null,
      "edges": Array [
        Object {
          "__typename": null,
          "cursor": "0",
          "node": Object {
            "__typename": null,
            "id": "0",
            "text": "nav1",
          },
        },
        Object {
          "__typename": null,
          "cursor": "1",
          "node": Object {
            "__typename": null,
            "id": "1",
            "text": "nav2",
          },
        },
      ],
      "pageInfo": Object {
        "__typename": null,
        "hasNextPage": false,
      },
    },
    "staredPosts": Array [
      Object {
        "__typename": "Post",
        "author": Object {
          "__typename": "User",
          "age": 10,
          "email": "wwwy3y3@gmail.com",
          "id": "1",
          "images": Array [
            Object {
              "url": "url",
            },
          ],
          "name": "user1",
          "posts": Array [
            "1",
            "2",
          ],
        },
        "id": "1",
        "notes": Object {
          "__typename": null,
          "edges": Array [
            Object {
              "__typename": null,
              "cursor": "0",
              "node": Object {
                "__typename": null,
                "id": "0",
                "text": "note1",
              },
            },
            Object {
              "__typename": null,
              "cursor": "1",
              "node": Object {
                "__typename": null,
                "id": "1",
                "text": "note2",
              },
            },
          ],
          "pageInfo": Object {
            "__typename": null,
            "hasNextPage": false,
          },
        },
        "title": "123",
      },
      Object {
        "__typename": "Post",
        "author": Object {
          "__typename": "User",
          "age": 20,
          "email": "wwwy3y3@gmail.com",
          "id": "2",
          "images": Array [
            Object {
              "url": "url",
            },
          ],
          "name": "user2",
          "posts": Array [
            "1",
            "2",
          ],
        },
        "id": "2",
        "notes": Object {
          "__typename": null,
          "edges": Array [
            Object {
              "__typename": null,
              "cursor": "0",
              "node": Object {
                "__typename": null,
                "id": "0",
                "text": "note3",
              },
            },
            Object {
              "__typename": null,
              "cursor": "1",
              "node": Object {
                "__typename": null,
                "id": "1",
                "text": "note4",
              },
            },
          ],
          "pageInfo": Object {
            "__typename": null,
            "hasNextPage": false,
          },
        },
        "title": "123",
      },
    ],
  },
}
`;
